Adds client hello section to usage guide (#3757)
aws/s2n-tls: docs/USAGE-GUIDE.md

Additionally, in TLS1.3, multiple session tickets may be issued for the same connection. Servers can call `s2n_config_set_initial_ticket_count()` to set the number of tickets they want to send and `s2n_connection_add_new_tickets_to_send()` to increase the number of tickets to send during a connection.

### Keying Material Lifetimes in Session Resumption

> Users may have questions on how long a secret generated from a full handshake is re-used with session resumption. The mechanism to expire old keying material varies based on the TLS version. In TLS1.2, the server only issues a new session ticket when doing a full handshake. Therefore the lifetime of the keying material inside the ticket is tied to the lifetime of the key used to encrypt the ticket. In TLS1.3, the encrypted ticket has an explicit lifetime stored in it, after which the keying material inside cannot be re-used. Servers can call `s2n_connection_set_server_keying_material_lifetime()` to configure this value.

Reviewer: ```suggestion
The lifetime of the secret used for session resumption varies based on the TLS version. In TLS1.2, the server only issues a new session ticket when doing a full handshake. Therefore the lifetime of the keying material inside the ticket is tied to the lifetime of the key used to encrypt the ticket. In TLS1.3, the encrypted ticket has an explicit lifetime stored in it, after which the keying material inside cannot be re-used. Servers can call `s2n_connection_set_server_keying_material_lifetime()` to configure this value.
```

Reviewer 2: A few changes to what Apoorv suggested. We have to be careful of the secret vs session ticket key confusion.
```suggestion
The lifetime of the secret, or "keying material", used for session resumption varies based on the TLS version. In TLS1.2, the server only issues a new session ticket when doing a full handshake. Therefore the lifetime of the keying material inside the ticket is tied to the lifetime of the session ticket key used to encrypt the ticket. In TLS1.3, the encrypted ticket has an explicit lifetime stored in it, after which the keying material inside cannot be re-used. Servers can call `s2n_connection_set_server_keying_material_lifetime()` to configure this value.
```

This might not even be enough-- we might need an explanation of what secret is reused in TLS1.2 vs TLS1.3 (the TLS1.2 one has no forward secrecy) and how that's different from the session ticket key.

## Client Hello Getters

> 

Reviewer 2: Idk if we should use "Getters" here. Maybe "Examining the Client Hello" or "Describing the Client Hello"

> s2n-tls stores the received Client Hello and makes it available to users for logging purposes or to branch off of its content. Call `s2n_connection_get_client_hello()` to get a pointer the `s2n_client_hello` struct storing the Client Hello. A NULL value will be returned if the connection has not yet received the Client Hello. The earliest point in the handshake when this struct is available is during the [Client Hello Callback](#client-hello-callback). 

Reviewer: ```suggestion
s2n-tls stores the received Client Hello and makes it available to the application for logging and etc. Call `s2n_connection_get_client_hello()` to get a pointer the `s2n_client_hello` struct storing the Client Hello. A NULL value will be returned if the connection has not yet received the Client Hello. The earliest point in the handshake when this struct is available is during the [Client Hello Callback](#client-hello-callback). 
```

Reviewer 2: ```suggestion
s2n-tls stores the received Client Hello and makes it available to the application. Call `s2n_connection_get_client_hello()` to get a pointer to the `s2n_client_hello` struct storing the Client Hello message. A NULL value will be returned if the connection has not yet received the Client Hello. The earliest point in the handshake when this struct is available is during the [Client Hello Callback](#client-hello-callback). 
```
I think we clean the client hello up if they free the handshake-- might be worth mentioning.

> Call `s2n_client_hello_get_raw_message()` to retrieve the complete Client Hello message with the random bytes on it zeroed out. Note that SSLv2 Client Hello messages are structured differently than other versions and thus their raw messages should be parsed accordingly (see [RFC5246](https://tools.ietf.org/html/rfc5246#appendix-E.2).) Call `s2n_connection_get_client_hello_version()` to retrieve the received Client Hello version.

Reviewer: ```suggestion
Call `s2n_client_hello_get_raw_message()` to retrieve the complete Client Hello message with the random bytes on it zeroed out. Note that SSLv2 Client Hello messages are structured differently than other versions and thus their raw messages should be parsed accordingly (see [RFC5246](https://tools.ietf.org/html/rfc5246#appendix-E.2).) Call `s2n_connection_get_client_hello_version()` to retrieve the Client Hello version.
```

`s2n_client_hello_get_cipher_suites()` will retrieve the list of cipher suites sent by the client.

`s2n_client_hello_get_session_id()` will get the session ID sent by the client in the ClientHello message. Note that this value may not be the session ID eventually associated with this particular connection since the session ID can change when the server sends the Server Hello. The official session ID can be retrieved with `s2n_connection_get_session_id()`.

Call `s2n_client_hello_get_extensions()` to retrieve the entire list of extensions sent in the Client Hello. Calling `s2n_client_hello_get_extension_by_id()` allows you to interrogate the `s2n_client_hello` struct for a specific extension.

### Client Hello Callback

> Users can gain access to the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()` with an implemented `s2n_client_hello_fn`. A common use-case for this is to switch the associated `s2n_config` based on information in the Client Hello. Note that `s2n_connection_server_name_extension_used()` MUST be invoked before exiting the callback if any of the connection properties were changed on the basis of the Server Name extension. If desired, the callback can return a negative value to make s2n-tls terminate the handshake early with a fatal handshake failure alert.

Reviewer: ```suggestion
Users can access the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()`. A common use-case for this is to switch the associated `s2n_config` based on information in the Client Hello. Note that `s2n_connection_server_name_extension_used()` MUST be invoked before exiting the callback if any of the connection properties were changed on the basis of the Server Name extension. If desired, the callback can return a negative value to terminate the handshake early with a fatal handshake failure alert.
```

> Users can gain access to the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()` with an implemented `s2n_client_hello_fn`. A common use-case for this is to switch the associated `s2n_config` based on information in the Client Hello. Note that `s2n_connection_server_name_extension_used()` MUST be invoked before exiting the callback if any of the connection properties were changed on the basis of the Server Name extension. If desired, the callback can return a negative value to make s2n-tls terminate the handshake early with a fatal handshake failure alert.

Reviewer 2: I wouldn't say "common usecase", maybe "A possible usecase" :) We also might want a warning here that modifying the connection or config mid-handshake based on the ClientHello is potentially dangerous and should be done carefully. In particular, switching from a config that supports TLS1.3 to one that doesn't opens the server up to a possible version downgrade attack. See https://github.com/aws/s2n-tls/issues/2223#issuecomment-686231386.

Maybe we also want to open an issue to record "could have done TLS1.3" before the client hello callback and base our downgrade protection off of that instead of off the current config after the callback.

#### Callback Modes

> The callback can be invoked in two modes: **S2N_CLIENT_HELLO_CB_BLOCKING**(default) and **S2N_CLIENT_HELLO_CB_NONBLOCKING**. Blocking mode will invoke the callback only once before the handshake continues. The handshake is therefore "blocked" until the callback returns a response code. Non-blocking mode will cause the handshake to pause at the Client Hello and return control back to the user. Further calls to `s2n_negotiate()` will return **S2N_FAILURE** with **S2N_ERR_T_BLOCKED** error type and **s2n_blocked_status** set to **S2N_BLOCKED_ON_APPLICATION_INPUT**. Only when the callback signals its work is complete by calling `s2n_client_hello_cb_done()` will the handshake continue.

Reviewer 3: Should `s2n_config_set_client_hello_cb_mode` be mentioned here?

> The callback can be invoked in two modes: **S2N_CLIENT_HELLO_CB_BLOCKING**(default) and **S2N_CLIENT_HELLO_CB_NONBLOCKING**. Use `s2n_config_set_client_hello_cb_mode()` to set the desired mode. Blocking mode will invoke the callback only once before the handshake continues. The handshake is therefore "blocked" until the callback returns a response code. Non-blocking mode will cause the handshake to pause at the Client Hello and return control back to the user. Further calls to `s2n_negotiate()` will return **S2N_FAILURE** with **S2N_ERR_T_BLOCKED** error type and **s2n_blocked_status** set to **S2N_BLOCKED_ON_APPLICATION_INPUT**. Only when the callback signals its work is complete by calling `s2n_client_hello_cb_done()` will the handshake continue. 

Reviewer 2: "will invoke the callback only once before the handshake continues". This is technically also what non-blocking mode does. Both modes only invoke the callback once.

Additionally, in TLS1.3, multiple session tickets may be issued for the same connection. Servers can call `s2n_config_set_initial_ticket_count()` to set the number of tickets they want to send and `s2n_connection_add_new_tickets_to_send()` to increase the number of tickets to send during a connection.

### Session Resumption Forward Secrecy

> In TLS1.2, the secret stored inside the ticket is the original session's master secret. Because of this, TLS1.2 session tickets are not forward secret, that is, compromising the session ticket key exposes both current and past sessions' encrypted data.
> 

Reviewer 2: Nit
```suggestion
In TLS1.2, the secret stored inside the ticket is the original session's master secret. Because of this, TLS1.2 session tickets are not forward secret, meaning that compromising the session ticket key exposes both current and past sessions' encrypted data.

```

Call `s2n_client_hello_get_raw_message()` to retrieve the complete Client Hello message with the random bytes on it zeroed out. Note that SSLv2 Client Hello messages are structured differently than other versions and thus their raw messages should be parsed accordingly (see [RFC5246](https://tools.ietf.org/html/rfc5246#appendix-E.2).) Call `s2n_connection_get_client_hello_version()` to retrieve the Client Hello version.

`s2n_client_hello_get_cipher_suites()` will retrieve the list of cipher suites sent by the client.

`s2n_client_hello_get_session_id()` will get the session ID sent by the client in the ClientHello message. Note that this value may not be the session ID eventually associated with this particular connection since the session ID can change when the server sends the Server Hello. The official session ID can be retrieved with `s2n_connection_get_session_id()`.

Call `s2n_client_hello_get_extensions()` to retrieve the entire list of extensions sent in the Client Hello. Calling `s2n_client_hello_get_extension_by_id()` allows you to interrogate the `s2n_client_hello` struct for a specific extension.

### Client Hello Callback

> Users can access the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()`. A possible use-case for this is to modify the `s2n_connection` based on information in the Client Hello. This should be done carefully as modifying the connection mid-handshake can be dangerous. In particular, switching from a `s2n_config` that supports TLS1.3 to one that does not opens the server up to a possible version downgrade attack. Note that `s2n_connection_server_name_extension_used()` MUST be invoked before exiting the callback if any of the connection properties were changed on the basis of the Server Name extension. If desired, the callback can return a negative value to make s2n-tls terminate the handshake early with a fatal handshake failure alert.
> 

Reviewer 2: Spacing nit
```suggestion
Users can access the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()`. A possible use-case for this is to modify the `s2n_connection` based on information in the Client Hello. This should be done carefully as modifying the connection mid-handshake can be dangerous. In particular, switching from a `s2n_config` that supports TLS1.3 to one that does not opens the server up to a possible version downgrade attack.

Note that `s2n_connection_server_name_extension_used()` MUST be invoked before exiting the callback if any of the connection properties were changed on the basis of the Server Name extension. If desired, the callback can return a negative value to make s2n-tls terminate the handshake early with a fatal handshake failure alert.
```

#### Callback Modes

> 
> The callback can be invoked in two modes: **S2N_CLIENT_HELLO_CB_BLOCKING** and **S2N_CLIENT_HELLO_CB_NONBLOCKING**. Use `s2n_config_set_client_hello_cb_mode()` to set the desired mode.

Reviewer 2: Nit
```suggestion

The callback can be invoked in two modes: **S2N_CLIENT_HELLO_CB_BLOCKING** or **S2N_CLIENT_HELLO_CB_NONBLOCKING**. Use `s2n_config_set_client_hello_cb_mode()` to set the desired mode.
```

The default mode, "blocking mode", will not return from the `s2n_negotiate()` call until the Client Hello callback returns. In this mode the callback should contain quick work like logging or simple configuration changes.
 

> In contrast, "non-blocking mode" will return from the `s2n_negotiate()` call if the Client Hello callback isn't finished yet, allowing the application to do work asynchronously. This mode should be used if the callback needs to make a network call or offload an expensive calculation to another thread. Only when the callback signals its work is complete by calling `s2n_client_hello_cb_done()` will the handshake continue.

Reviewer 2: Not quite. Technically the s2n_negotiate call still returns after the Client Hello callback finishes, it just doesn't continue the handshake until s2n_client_hello_cb_done is called. The callback still has to return immediately.

## Examining the Client Hello

s2n-tls stores the received Client Hello and makes it available to the application. Call `s2n_connection_get_client_hello()` to get a pointer to the `s2n_client_hello` struct storing the Client Hello message. A NULL value will be returned if the connection has not yet received the Client Hello. The earliest point in the handshake when this struct is available is during the [Client Hello Callback](#client-hello-callback). The stored Client Hello message will not be available after calling `s2n_connection_free_handshake()`.

Call `s2n_client_hello_get_raw_message()` to retrieve the complete Client Hello message with the random bytes on it zeroed out. Note that SSLv2 Client Hello messages are structured differently than other versions and thus their raw messages should be parsed accordingly (see [RFC5246](https://tools.ietf.org/html/rfc5246#appendix-E.2).) Call `s2n_connection_get_client_hello_version()` to retrieve the Client Hello version.

`s2n_client_hello_get_cipher_suites()` will retrieve the list of cipher suites sent by the client.

> `s2n_client_hello_get_session_id()` will get the session ID sent by the client in the ClientHello message. Note that this value may not be the session ID eventually associated with this particular connection since the session ID can change when the server sends the Server Hello. The official session ID can be retrieved with `s2n_connection_get_session_id()`.
> 

Reviewer 2: ```suggestion
`s2n_client_hello_get_session_id()` will get the session ID sent by the client in the ClientHello message. Note that this value may not be the session ID eventually associated with this particular connection since the session ID can change when the server sends the Server Hello. The official session ID can be retrieved with `s2n_connection_get_session_id()` after the handshake completes.

```

Call `s2n_client_hello_get_extensions()` to retrieve the entire list of extensions sent in the Client Hello. Calling `s2n_client_hello_get_extension_by_id()` allows you to interrogate the `s2n_client_hello` struct for a specific extension.

### Client Hello Callback

> Users can access the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()`. A possible use-case for this is to modify the `s2n_connection` based on information in the Client Hello. This should be done carefully as modifying the connection mid-handshake can be dangerous. In particular, switching from a `s2n_config` that supports TLS1.3 to one that does not opens the server up to a possible version downgrade attack. 
> 

Reviewer 2: I'm now wondering if "dangerous" is too harsh. Most things you do are safe, there are just things you can do that are unsafe. Maybe if we're more specific about what exactly is dangerous?
```suggestion
Users can access the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()`. A possible use-case for this is to modify the `s2n_connection` based on information in the Client Hello. This should be done carefully, as modifying the connection in response to untrusted input can be dangerous. In particular, switching from an `s2n_config` that supports TLS1.3 to one that does not opens the server up to a possible version downgrade attack. 

```

`s2n_connection_server_name_extension_used()` MUST be invoked before exiting the callback if any of the connection properties were changed on the basis of the Server Name extension. If desired, the callback can return a negative value to make s2n-tls terminate the handshake early with a fatal handshake failure alert.

#### Callback Modes

The callback can be invoked in two modes: **S2N_CLIENT_HELLO_CB_BLOCKING** or **S2N_CLIENT_HELLO_CB_NONBLOCKING**. Use `s2n_config_set_client_hello_cb_mode()` to set the desired mode.

> The default mode, "blocking mode", will not return from the `s2n_negotiate()` call until the Client Hello callback returns. In this mode the callback should contain quick work like logging or simple configuration changes.
>  
> In contrast, "non-blocking mode" will return from the `s2n_negotiate()` call even if the Client Hello callback hasn't been marked as done yet. This will pause the handshake and allow the application to do work asynchronously. Only when the callback signals its work is complete by calling `s2n_client_hello_cb_done()` will the handshake continue. This mode should be used if the callback needs to make a network call or offload an expensive calculation to another thread. 

Reviewer 2: I'm still not quite sure this communicates what we're trying to communicate :\

Maybe:
```suggestion
The default mode, "blocking mode", will wait for the ClientHello callback to succeed and then continue the handshake. Use this mode for light-weight callbacks that won't slow down the handshake or block the main thread, like logging or simple configuration changes.

In contrast, "non-blocking mode" will wait for the ClientHello callback to succeed and then pause the handshake, immediately returning from s2n_negotiate with an error indicating that the handshake is blocked on application input. This allows the application to do expensive or time-consuming work like network calls without blocking the main thread. Only when the application calls `s2n_client_hello_cb_done()` will the handshake be able to resume.
```
